lib/active_record/connection_adapters/sqlserver/database_statements.rb

```diff
@@ -206,6 +206,35 @@ def build_sql_for_merge_insert(insert:, insert_all:, columns_with_uniqueness_con
           sql
         end
 
+        # For `nil` identity columns we need to ensure that the values do not match so that they are all inserted.
+        # Method is a combination of `ActiveRecord::InsertAll#values_list` and `ActiveRecord::ConnectionAdapters::SQLServer::DatabaseStatements#default_insert_value`.
+        def merge_insert_values_list(insert:, insert_all:)
+          connection = insert.send(:connection)
+          identity_index = 0
+
+          types = insert.send(:extract_types_from_columns_on, insert.model.table_name, keys: insert.keys_including_timestamps)
+
+          values_list = insert_all.map_key_with_value do |key, value|
+            if Arel::Nodes::SqlLiteral === value
+              value
+            elsif insert.primary_keys.include?(key) && value.nil?
+              column = insert.send(:column_from_key, key)
+
+              if column.is_identity?
```

Comment:
just curious: is it possible under MSSQL that you have an identity column that is not a primary key? or is identity column just another word for "column with autogenerated identifier"?

Reply:
Had to look that up. Looks like you can have a primary key that is not an identity column and an identity column that is not a primary key.

```sql
CREATE TABLE ExampleTable (
    PrimaryKeyCol INT NOT NULL PRIMARY KEY,
    IdentityColumn INT IDENTITY(1,1) NOT NULL,
    SomeOtherCol NVARCHAR(100) NULL
);
```